pipeline/metadata/add_metadata.py

```diff
@@ -75,17 +109,96 @@ def annotate_row_ip(
 
     return rows_with_metadata
 
+  def annotate_answer_ips(
+      self, rows: beam.pvalue.PCollection[SatelliteRow]
+  ) -> beam.pvalue.PCollection[SatelliteRow]:
+    """Add ip metadata info to received ip lists in rows
+
+    Args:
+      rows: PCollection of SatelliteRows
+
+    Returns:
+      SatelliteRows with routeview metadata attatched to answer ips.
+    """
+    # PCollection[Tuple[roundtrip_id, SatelliteRow]]
+    rows_with_roundtrip_id = (
+        rows | 'add roundtrip_ids: answer metadata' >>
+        beam.Map(set_random_roundtrip_id).with_output_types(
+            Tuple[str, SatelliteRow]))
+
+    # PCollection[Tuple[DateIpKey, Tuple[roundtrip_id, SatelliteAnswer]]]
+    received_ips_keyed_by_ip_and_date = (
+        rows_with_roundtrip_id |
+        'get received ips: answer metadata' >> beam.FlatMap(
+            _get_received_ips_with_roundtrip_id_and_date).with_output_types(
+                Tuple[DateIpKey, Tuple[str, SatelliteAnswer]]))
+
+    #PCollection[DateIpKey]
+    # pylint: disable=no-value-for-parameter
+    ips_and_dates = (
+        received_ips_keyed_by_ip_and_date | 'get ip and date keys per answer' >>
+        beam.Keys().with_output_types(DateIpKey))
+
+    # PCollection[DateIpKey]
+    deduped_ips_and_dates = (
+        # pylint: disable=no-value-for-parameter
+        ips_and_dates |
+        'dedup answer ips' >> beam.Distinct().with_output_types(DateIpKey))
+
+    # PCollection[Tuple[date,List[ip]]]
+    grouped_ips_by_dates = (
+        deduped_ips_and_dates | 'group answer ips by date' >>
+        beam.GroupByKey().with_output_types(Tuple[str, Iterable[str]]))
+
+    # PCollection[Tuple[DateIpKey,IpMetadataWithDateKey]]
+    ips_with_metadata = (
+        grouped_ips_by_dates | 'get ip metadata for answers' >>
+        beam.FlatMapTuple(self._annotate_ips).with_output_types(
+            Tuple[DateIpKey, IpMetadataWithDateKey]))
+
+    grouped_metadata_and_received_ips = (({
+        IP_METADATA_PCOLLECTION_NAME: ips_with_metadata,
+        RECEIVED_IPS_PCOLLECTION_NAME: received_ips_keyed_by_ip_and_date
+    }) | 'add received ip metadata: group by keys' >> beam.CoGroupByKey())
+
+    # PCollection[Tuple[roundtrip_id, SatelliteAnswerWithDateKey]] received ip row with
+    received_ips_with_metadata = (
+        grouped_metadata_and_received_ips |
+        'add received ip metadata: merge tags with answers' >> beam.
+        FlatMapTuple(merge_satellite_metadata_with_answers).with_output_types(
+            Tuple[str, SatelliteAnswerWithDateKey]))
+
+    # PCollection[Tuple[roundtrip_id, List[Tuple[roundtrip_id, SatelliteAnswerWithDateKey]]]]
+    received_ips_grouped_by_roundtrip_ip = (
+        received_ips_with_metadata |
+        'group received_by roundtrip: answer metadata' >>
+        beam.GroupBy(lambda x: x[0]).with_output_types(
```

Comment:
Is this assuming that there's only one value or are you picking the first? Please add a comment

Reply:
This is grouping by the first element of a 2-tuple. It's to get all the answers for a given roundtrip back together.

PCollection[Tuple[roundtrip_id, SatelliteAnswerWithDateKey]]
 -> group by roundtrip_id
PCollection[Tuple[roundtrip_id, List[Tuple[roundtrip_id, SatelliteAnswerWithDateKey]]]]

Cleaned up the existing comments and changed the lambda to
`lambda kv: kv[0]`
to hopefully make it clearer that we're getting the key.